webui/react/src/utils/sort.ts

```diff
@@ -0,0 +1,93 @@
+import { CommandState, MetricName, MetricType, Primitive, RunState, State } from 'types';
+
+const runStateSortValues: Record<RunState, number> = {
+  [RunState.Active]: 0,
+  [RunState.Paused]: 1,
+  [RunState.StoppingError]: 2,
+  [RunState.Errored]: 3,
+  [RunState.StoppingCompleted]: 4,
+  [RunState.Completed]: 5,
+  [RunState.StoppingCanceled]: 6,
+  [RunState.Canceled]: 7,
+  [RunState.Deleted]: 7,
+  [RunState.Unspecified]: 8,
+};
+
+const commandStateSortValues: Record<CommandState, number> = {
+  [CommandState.Pending]: 0,
+  [CommandState.Assigned]: 1,
+  [CommandState.Pulling]: 2,
+  [CommandState.Starting]: 3,
+  [CommandState.Running]: 4,
+  [CommandState.Terminating]: 5,
+  [CommandState.Terminated]: 6,
+};
+
+export const alphanumericSorter = (a: string | number, b: string | number): number => {
+  return a.toString().localeCompare(b.toString(), 'en', { numeric: true });
+};
+
+export const booleanSorter = (a: boolean, b: boolean): number => {
+  // True values first.
+  return (a === b) ? 0 : (a ? -1 : 1);
+};
+
+export const commandStateSorter = (a: CommandState, b: CommandState): number => {
+  return commandStateSortValues[a] - commandStateSortValues[b];
+};
+
+/*
+ * Sort the metric names by having the validation metrics come first followed by training metrics.
+ * Within each type of metric, sort in the order they appear in the `MetricNames` array.
+ * Within the respective type of metrics, `MetricNames` is currently sorted alphanumerically.
+ */
+export const metricNameSorter = (a: MetricName, b: MetricName): number => {
+  const isAValidation = a.type === MetricType.Validation;
+  const isBValidation = b.type === MetricType.Validation;
+  if (isAValidation && !isBValidation) return -1;
+  if (isBValidation && !isAValidation) return 1;
+  return alphanumericSorter(a.name, b.name);
+};
+
+export const numericSorter = (a?: number, b?: number, reverseOrder = false): number => {
+  if (a != null && b != null) {
+    const diff = reverseOrder ? b - a : a - b;
+    if (diff < 0) return -1;
+    if (diff > 0) return 1;
+    return 0;
+  }
+  if (a != null && b == null) return reverseOrder ? -1 : 1;
+  if (a == null && b != null) return reverseOrder ? 1 : -1;
+  return 0;
+};
+
+export const primitiveSorter = (a: Primitive, b: Primitive): number => {
+  if (typeof a === 'boolean' && typeof b === 'boolean') {
+    return booleanSorter(a, b);
+  } else if (typeof a === 'number' && typeof b === 'number') {
+    return numericSorter(a, b);
+  } else if (typeof a === 'string' && typeof b === 'string') {
+    return alphanumericSorter(a, b);
+  }
+  return 0;
+};
+
+export const runStateSorter = (a: RunState, b: RunState): number => {
+  return runStateSortValues[a] - runStateSortValues[b];
+};
+
+export const stringTimeSorter = (a: string, b: string): number => {
```

Comment:
Is that one really needed? since time strings (eg: `2021-02-24T07:53:06.911722Z`) can simply be sorted alphabetically?

Reply:
good question. Not sure what circumstances were, but I remember Hamid's and my respective local clusters reported different strings. Where his had precision to the microseconds where mine only had milliseconds. `2021-02-24T07:53:06.911722Z` vs `2021-02-24T07:53:06.911Z`. I also don't know if we can completely rely on timezones to always be UTC. So in the case where we are comparing different timezones and different precisions, converting them to date objects prior to comparison might be safer.

Reply:
Thanks for the explanation.